### DISCUSSION TITLE
How to use roslynator packages for custom analysers

### DISCUSSION THREAD

**Discussion by nathanpovo** [JosefPihrt/Roslynator#1213] [Q&A]
How possible it is to use the roslynator packages to write your own custom packages?

The docs mention that some of the packages are available for developers to be able to write and test their custom analysers

https://github.com/dotnet/roslynator/blob/32c1804bbdef8b832f18debb59072bd1645d34fb/README.md?plain=1#L50-L63

However, I could not find out how to actually use them. The documentation does not go into detail on this.

The unit tests in the solution use a number of internal types (making it impossible for me to use them in a custom roslyn analyser). There are also a number of auto-generated c# classes in this repo (that are referenced by the unit tests), but there is no info on how these were generated.

**Comment by josefpihrt** [comment]
Hi,

Thank you for the question. I would separate this issue into two parts:

### .NET API extending Roslyn API (packages Roslynator.Core etc.)

I added [this page](https://josefpihrt.github.io/docs/roslynator/how-to-use-net-api) about how to use .NET API in a custom analyzer package.


### Usage of testing framework

I prepared [PR](https://github.com/dotnet/roslynator/pull/1214) that makes lot of types public which makes the development of unit test easier.

And here is [PR](https://github.com/josefpihrt/josefpihrt.github.io/pull/28) that updates docs.

Any comments of the PRs are welcome.

Nevertheless, technically this PR introduces a breaking change which mean that it should be merged when next major version is released which is usually when next major version of Roslyn is released. I have to think about it whether to merge it earlier.

The good news is that there very simple workaround: just copy project "Tests.Common" to your project and use it. Then the development experience will be virtually the same as with the PR merged.

**Reply by nathanpovo** [reply]
Thank you for getting back to me so quickly!

Regarding this section of the docs https://josefpihrt.github.io/docs/roslynator/how-to-use-net-api#analyzer-package-configuration, is that only relevant for when the analyser is being packaged for nuget? I wouldn't need to worry about this while developing locally, right?

**Reply by josefpihrt** [reply]
Correct.

**Comment by nathanpovo** [comment]
> There are also a number of auto-generated c# classes in this repo (that are referenced by the unit tests), but there is no info on how these were generated.

Any info on this?

These are the files I am referring to

https://github.com/dotnet/roslynator/blob/32c1804bbdef8b832f18debb59072bd1645d34fb/src/Analyzers/CSharp/DiagnosticRules.Generated.cs

https://github.com/dotnet/roslynator/blob/32c1804bbdef8b832f18debb59072bd1645d34fb/src/Analyzers/CSharp/DiagnosticIdentifiers.Generated.cs

The (new) docs on unit testing also reference them

https://github.com/josefpihrt/josefpihrt.github.io/blob/f39502ef76fd6a915f5914befc1cf2c8b0ef0bed/docs/roslynator/analyzers-testing.md?plain=1#L26

**Reply by josefpihrt** [reply]
It' not important that the classes are auto-generated.

Important fact is that when you want to test some analyzer you have to include DiagnosticDescriptor in your unit test class. It's up you where you define DiagnosticDescriptor for your analyzer.

DiagnosticIdentifiers contains just list of diagnostic IDs it's not necessary to include them in the unit test. In fact, diagnostic ID is part of DiagnosticDescriptor.

> The (new) docs on unit testing also reference them

I will remove it to avoid any further confusion.

**Comment by josefpihrt** [comment]
@nathanpovo Roslynator 4.6.0 has been released. It contains changes from #1214

Documentation has been updated accordingly https://josefpihrt.github.io/docs/roslynator/testing
